storage.go

```diff
@@ -242,7 +242,10 @@ func (ms *MemoryStorage) Compact(compactIndex uint64) error {
 	}
 
 	i := compactIndex - offset
-	ents := make([]pb.Entry, 1, 1+uint64(len(ms.ents))-i)
+	// NB: allocate a new slice instead of reusing the old ms.ents. Entries in
+	// ms.ents are immutable, and can be referenced from outside MemoryStorage
+	// through slices returned by ms.Entries().
+	ents := make([]pb.Entry, 1, uint64(len(ms.ents))-i)
 	ents[0].Index = ms.ents[i].Index
 	ents[0].Term = ms.ents[i].Term
 	ents = append(ents, ms.ents[i+1:]...)
```

Comment:
The meta question is, why optimize this testing-only implementation?
One way to do so would be to allow a more streamlined `raft.Storage` implementation, where instead of a user re-implementing a bunch of logic every time, they would just plug their storage engine (with a much narrower API) into a struct that we provide (essentially `MemoryStorage` but wrapping this user-supplied object rather than a slice).

Reply:
> The meta question is, why optimize this testing-only implementation?


It isn't test-only. It's used by etcd. Please see [server.go#L2114](https://github.com/etcd-io/etcd/blob/29e79717217f49c46654309b5fb74546f7062967/server/etcdserver/server.go#L2114)

> allow a more streamlined raft.Storage implementation, where instead of a user re-implementing a bunch of logic every time, they would just plug their storage engine (with a much narrower API) into a struct that we provide

It's a big topic, which definitely should be discussed separately in a dedicated session.

Reply:
> Couldn't get the idea. What narrower API are you thinking of?

Not worth discussing further here, but each user of raft needs to implement `raft.Storage` and thus needs to re-implement a bunch of subtle details regarding how `Entries()` has to work, etc.

Instead, most users probably have something that can write an entry and point read an entry, so unless they want something more bespoke that thing that they have should be enough to get a `raft.Storage` implementation in which all of the subtlety is provided by raft.